File: examples/strapi-content-feed/tests/pipeline.test.ts
import { Pipeline } from '../src/pipeline/runner';
import { Source, Hydrator, Filter, Scorer, Selector, SideEffect, ScoredItem } from '../src/pipeline/interfaces';

interface TestCtx { userId: number; }
interface TestItem { id: number; authorId: number; title: string; }

describe('Pipeline', () => {
  it('runs the six stages in order', async () => {
    const callOrder: string[] = [];

    const source: Source<TestCtx, TestItem> = {
      name: 'src',
      async fetch() {
        callOrder.push('source');
        return [
          { id: 1, authorId: 10, title: 'a' },
          { id: 2, authorId: 10, title: 'b' },
          { id: 3, authorId: 20, title: 'c' },
        ];
      },
    };

    const hydrator: Hydrator<TestCtx, TestItem> = {
      name: 'hyd',
      async hydrate() { callOrder.push('hydrator'); },
    };

    const filter: Filter<TestCtx, TestItem> = {
      name: 'fil',
      keep(_ctx, item) {
        callOrder.push(`filter:${item.id}`);
        return item.id !== 2;
      },
    };

    const scorer: Scorer<TestCtx, TestItem> = {
      name: 'sc',
      async score(_ctx, items) {
        callOrder.push('scorer');
        for (const s of items) s.score = s.item.id;
      },
    };

    const selector: Selector<TestCtx, TestItem> = {
      name: 'sel',
      select(_ctx, items, k) {
        callOrder.push('selector');
        return [...items].sort((a, b) => b.score - a.score).slice(0, k);
      },
    };

    const sideEffect: SideEffect<TestCtx, TestItem> = {
      name: 'se',
      emit() { callOrder.push('side_effect'); },
    };

    const pipeline = new Pipeline<TestCtx, TestItem>({
      sources: [source],
      hydrators: [hydrator],
      filters: [filter],
      scorers: [scorer],
      selector,
      sideEffects: [sideEffect],
    });

    const result = await pipeline.run({ userId: 1 }, 10);

    // Filter dropped id=2
    expect(result.map(r => r.item.id)).toEqual([3, 1]);

    // Stage order
    const stageOrder = callOrder.filter(s => !s.startsWith('filter:'));
    expect(stageOrder).toEqual(['source', 'hydrator', 'scorer', 'selector', 'side_effect']);
  });

  it('runs sources in parallel', async () => {
    const sourceTimings: { name: string; t: number }[] = [];

    const makeSource = (name: string, delay: number): Source<TestCtx, TestItem> => ({
      name,
      async fetch() {
        const start = Date.now();
        await new Promise(r => setTimeout(r, delay));
        sourceTimings.push({ name, t: Date.now() - start });
        return [{ id: 1, authorId: 10, title: name }];
      },
    });

    const pipeline = new Pipeline<TestCtx, TestItem>({
      sources: [
        makeSource('s1', 50),
        makeSource('s2', 50),
        makeSource('s3', 50),
      ],
      hydrators: [],
      filters: [],
      scorers: [],
      selector: {
        name: 'sel',
        select: (_ctx, items, k) => items.slice(0, k),
      },
      sideEffects: [],
    });

    const start = Date.now();
    await pipeline.run({ userId: 1 }, 10);
    const total = Date.now() - start;

    // Sequential would take ~150ms; parallel ~50ms. Allow generous margin.
    expect(total).toBeLessThan(120);
  });

  it('survives a source error', async () => {
    const goodSource: Source<TestCtx, TestItem> = {
      name: 'good',
      async fetch() { return [{ id: 1, authorId: 10, title: 'a' }]; },
    };

    const badSource: Source<TestCtx, TestItem> = {
      name: 'bad',
      async fetch() { throw new Error('source down'); },
    };

    const pipeline = new Pipeline<TestCtx, TestItem>({
      sources: [goodSource, badSource],
      hydrators: [],
      filters: [],
      scorers: [],
      selector: {
        name: 'sel',
        select: (_ctx, items, k) => items.slice(0, k),
      },
      sideEffects: [],
    });

    const result = await pipeline.run({ userId: 1 }, 10);
    expect(result.length).toBe(1);
    expect(result[0].item.id).toBe(1);
  });
});

File: examples/zentra-go/pipeline/pipeline_test.go
package pipeline

import (
	"context"
	"errors"
	"testing"
	"time"
)

type testCtx struct{ UserID int }
type testItem struct {
	ID       int
	AuthorID int
	Title    string
}

type fakeSource struct {
	name  string
	delay time.Duration
	items []testItem
	err   error
}

func (s *fakeSource) Name() string { return s.name }
func (s *fakeSource) Fetch(ctx context.Context, c testCtx) ([]testItem, error) {
	if s.delay > 0 {
		time.Sleep(s.delay)
	}
	return s.items, s.err
}

type evenFilter struct{}

func (evenFilter) Name() string                                          { return "even" }
func (evenFilter) Keep(_ context.Context, _ testCtx, item testItem) bool { return item.ID%2 == 0 }

type idScorer struct{}

func (idScorer) Name() string { return "id" }
func (idScorer) Score(_ context.Context, _ testCtx, items []ScoredItem[testItem]) error {
	for i := range items {
		items[i].Score = float64(items[i].Item.ID)
	}
	return nil
}

type topKSelector struct{}

func (topKSelector) Name() string { return "topk" }
func (topKSelector) Select(_ testCtx, items []ScoredItem[testItem], k int) []ScoredItem[testItem] {
	SortByScoreDesc(items)
	if k > len(items) {
		k = len(items)
	}
	return items[:k]
}

func TestPipelineFiltersAndSorts(t *testing.T) {
	p := &Pipeline[testCtx, testItem]{
		Sources: []Source[testCtx, testItem]{
			&fakeSource{
				name: "s1",
				items: []testItem{
					{ID: 1, AuthorID: 10, Title: "a"},
					{ID: 2, AuthorID: 10, Title: "b"},
					{ID: 3, AuthorID: 20, Title: "c"},
					{ID: 4, AuthorID: 20, Title: "d"},
				},
			},
		},
		Filters:  []Filter[testCtx, testItem]{evenFilter{}},
		Scorers:  []Scorer[testCtx, testItem]{idScorer{}},
		Selector: topKSelector{},
	}

	result, err := p.Run(context.Background(), testCtx{UserID: 1}, 5)
	if err != nil {
		t.Fatalf("Run failed: %v", err)
	}

	if len(result) != 2 {
		t.Fatalf("expected 2 results (even IDs), got %d", len(result))
	}
	if result[0].Item.ID != 4 || result[1].Item.ID != 2 {
		t.Errorf("expected [4, 2], got [%d, %d]", result[0].Item.ID, result[1].Item.ID)
	}
}

func TestPipelineSourcesParallel(t *testing.T) {
	p := &Pipeline[testCtx, testItem]{
		Sources: []Source[testCtx, testItem]{
			&fakeSource{name: "s1", delay: 50 * time.Millisecond, items: []testItem{{ID: 1}}},
			&fakeSource{name: "s2", delay: 50 * time.Millisecond, items: []testItem{{ID: 2}}},
			&fakeSource{name: "s3", delay: 50 * time.Millisecond, items: []testItem{{ID: 3}}},
		},
		Selector: topKSelector{},
	}

	start := time.Now()
	_, err := p.Run(context.Background(), testCtx{UserID: 1}, 10)
	elapsed := time.Since(start)

	if err != nil {
		t.Fatal(err)
	}
	// Sequential: ~150ms. Parallel: ~50ms. Generous bound.
	if elapsed > 120*time.Millisecond {
		t.Errorf("sources appear sequential: took %v", elapsed)
	}
}

func TestPipelineSurvivesSourceError(t *testing.T) {
	p := &Pipeline[testCtx, testItem]{
		Sources: []Source[testCtx, testItem]{
			&fakeSource{name: "good", items: []testItem{{ID: 1}}},
			&fakeSource{name: "bad", err: errors.New("boom")},
		},
		Selector: topKSelector{},
	}

	result, err := p.Run(context.Background(), testCtx{UserID: 1}, 10)
	if err != nil {
		t.Fatal(err)
	}
	if len(result) != 1 || result[0].Item.ID != 1 {
		t.Errorf("expected single item ID=1, got %v", result)
	}
}
